### PR TITLE
feat: new sharpe ratio formula

### PR DESCRIPTION
**Pull request by xzmeng**

### CODE REVIEW

File: freqtrade/data/metrics.py [hunk 1]
```diff
@@ -313,31 +317,64 @@ def calculate_sortino(
...
+                df.loc[filled_at:, stake_currency] += stake - fee
+
+    for pair in pairlist:
+        close_price = btdata[pair].set_index("date")["close"]
+        df[pair] = df[pair] * close_price
+    value_in_stake_currency = df.sum(axis=1)
+    return value_in_stake_currency
+
+
+def calculate_sharpe(account_value_history: pd.Series, timeframe: str) -> float:
```
Comment (xmatthias) [inline freqtrade/data/metrics.py:362]: while i don't necessarily mind having a "per account" sharpe ratio - it shouldn't hard-replace the existing one - at least not initially.

File: freqtrade/data/metrics.py [hunk 2]
```diff
@@ -313,31 +317,64 @@ def calculate_sortino(
...
+        for order in trade.orders:
+            filled_at = pd.Timestamp(dt_from_ts(order["order_filled_timestamp"]))
+            stake = order["safe_price"] * order["amount"]
+
+            if order["ft_is_entry"]:
+                fee = stake * trade.fee_open
+            else:
+                fee = stake * trade.fee_close
+
+            if order["ft_order_side"] == "buy":
```
Comment (xmatthias) [inline freqtrade/data/metrics.py:348]: wrong for shorts ?

File: freqtrade/data/metrics.py [hunk 3]
```diff
@@ -313,31 +317,64 @@ def calculate_sortino(
...
+    every candle stick, in stake currency.
+    """
+    index = pd.date_range(min_date, max_date, freq=timeframe_to_resample_freq(timeframe))
+    df = pd.DataFrame(index=index)
+    df[stake_currency] = float(start_balance)
+    df[pairlist] = 0.0
+    for trade in trades.itertuples():
+        for order in trade.orders:
+            filled_at = pd.Timestamp(dt_from_ts(order["order_filled_timestamp"]))
+            stake = order["safe_price"] * order["amount"]
```
Comment (xmatthias) [inline freqtrade/data/metrics.py:341]: i'm not sure that's actually correct (not for what you aim to do, anyway).

My assumption (based on the lack of description) is that this refers to the discussion on discord - which in my understanding - would create a dataframe of "timerange" lenght with interval "timeframe" - and with the account value at every step of the way.

Now let's assume a trade of 4h length (1h candles)
You can use safe_price * amount for the first candle of that trade to get the initial amount
latest on the 2nd candle, we should however use one of the available prices (close/open) - as the account value (open and closed trades) did change - so the account may be 10% higher - or 10% lower. By using safe_price - it's going to be 0% - independent of actual price movements.

Comment (xzmeng) [reply freqtrade/data/metrics.py:341]: The first step is to calculate the  amount of every currencies in our account，so the same rule should apply to the quote currency. The second step - multiply every currency by its open/close price and then sum them up is where the value change take effect.

For convinience，I used the pair name as the column name，they are actually the amount of the base currency of the pair.

Comment (xmatthias) [reply freqtrade/data/metrics.py:341]: yeah i noticed - the usage is essentially explained further below 👍

**Comment by xmatthias** [comment]
@xzmeng did this work / not work / not worth pursuing?

i find especially the "account value" function of great interest - and a parallel sharpe function can also help (i'd want to compare them, at least initially)

**Comment by xzmeng** [comment]
it works for spot IMO, but I haven't explored futures mode yet.

**Comment by Kritoooo** [comment]
Hi @xmatthias and @xzmeng,
I’d like to continue the work started here. The core idea—an account-value-based Sharpe ratio—still seems valuable, because I’ve found that the same calculate_account_value_history function is also needed for the metrics mentioned in #11733 .
I’ll address the outstanding review points:
1. Keep the new Sharpe function separate so the legacy one remains untouched.
2. Fix the short-trades handling in calculate_account_value_history (amount sign & fee direction).

If you have time, would you be willing to finish this PR?
If you’re busy, I’m happy to pick it up—just let me know whether I should open a fresh PR that references this discussion or push commits directly to the existing branch.

**Comment by xmatthias** [comment]
neither, really. 
I've got a (better) version of a similar approach in the works - one that works for a wider usecase and more cases (e.g. futures) - but is based on this - both in idea and code.

This Pr (or approach would also not work for #11733 - as you can't take the assumption that "data will be available". The current PR here is only applicable for backtesting/hyperopt - while the issue you link to is for live.


I'm currently still leaving this open - but neither with the plan to finish this PR, nor to have someone else pick up and finish it.
When time allows, i'll try to finish my implementation - but until then - work on this is pretty much pointless as it'd replicate my work and - unless it provides more than what mine can do already - wouldn't be accepted.

**Comment by Kritoooo** [comment]
Thanks for the heads-up! I hadn hadn’t realised you already had a more comprehensive implementation in progress.
I’ll stop working on this branch so we don’t duplicate effort. Looking forward to the release of your implementation!

**Comment by itsrayeedscodespace** [comment]
Given that @xmatthias already has a more complete and futures-compatible implementation in progress, it seems more productive to avoid duplicating effort here. Still, the work done in this PR helped surface several useful edge cases (particularly around shorts and account-value tracking), which should be valuable for the final implementation.

Happy to help evaluate or test the upcoming version once it's ready — especially in mixed spot/futures setups and with different fee models.

**Comment by xmatthias** [comment]
I'm going to close this here as this solution was incomplete - though it served as inspiration for the new wallet balance functionality.

The latest develop does however include the sharpe ratio based on unrealized profits - as well as a couple of other metrics that follow the same pattern.
